provider migrate cloudant

o-nikolas opened the pull request:
Migrates the Cloudant provider.

static and unit tests pass. 

This provider required changes to the new provider schema to add back support for excluded python versions which cloudant makes use of (see comment annotations for more details).  The release management scripts for generating project toml files also needed updating.

related: https://github.com/apache/airflow/issues/46045

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

o-nikolas committed:
Update provider schema to include excluded-python-versions

The provider schema previously supported this field but it was dropped
during the provider migration project. Cloudant is the only provider at
the time of writing which makes use of this mechanism

o-nikolas committed:
Migrate Cloudant provider to new structure

o-nikolas commented on airflow/new_provider.yaml.schema.json:
```diff
@@ -31,6 +31,13 @@
                 "removed"
             ]
         },
+        "excluded-python-versions": {
```
Return support for excluded python versions. Copied from the old schema

o-nikolas commented on dev/breeze/src/airflow_breeze/prepare_providers/provider_documentation.py:
```diff
@@ -1222,6 +1222,13 @@ def _regenerate_pyproject_toml(context: dict[str, Any], provider_details: Provid
+    for excluded_python_version in provider_details.excluded_python_versions:
+        requires_python_version += f", !={excluded_python_version}"
+
+    context["REQUIRES_PYTHON"] = requires_python_version
```
Dynamically generate the requires-python toml field now that it can have exclusions.

o-nikolas commented on providers/cloudant/pyproject.toml:
```diff
@@ -0,0 +1,79 @@
+    "Programming Language :: Python :: 3.12",
+    "Topic :: System :: Monitoring",
+]
+requires-python = "~=3.9, !=3.9"
```
@potiuk can you double check this one? Is the exclusion correct? I suspect `!=3.9` isn't legal and it should maybe be `!=3.9.*` instead?

Also can this string be comma delimited like this? I found very little documentation and some of it is conflicting.

potiuk replied:
Specification is here: https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers amd


And yes https://packaging.python.org/en/latest/specifications/version-specifiers/#version-exclusion !=3.9.* sounds right

o-nikolas replied:
> And yes https://packaging.python.org/en/latest/specifications/version-specifiers/#version-exclusion !=3.9.* sounds right

Okay, I propose that we update the provider.yaml file to specify `3.9.*` then, rather than putting in brittle code here to detect that no wildcard was provided. It will set the example for future providers if they want to do that as well.

o-nikolas replied:
Turns out this breaks some other automation we have, so this will need more work. I may pivot to adding the `.*` if it's missing.

I am also curious if it will just work how it is now. I'm not sure the steps to verify that

o-nikolas replied:
When the requires-python field is updated to be `"~=3.9,!=3.9.*"` the airflow 3.9 build step fails (error below), which makes sense because no resolution is possible using python 3.9. I'm actually curious how this was ever passing before...

Any thoughts or insight on this one @potiuk?


```
   #45 0.956 Installing all packages in eager upgrade mode. Installation method: .
  #45 0.956 
  #45 0.956 + uv pip install --system --upgrade --resolution highest --editable '.[devel-ci]' --editable ./task_sdk --editable providers/pagerduty --editable providers/common/compat --editable providers/common/io --editable providers/common/sql --editable providers/cloudant --editable providers/asana --editable providers/pinecone --editable providers/cohere --editable providers/grpc --editable providers/singularity --editable providers/elasticsearch --editable providers/docker --editable providers/vertica --editable providers/snowflake --editable providers/standard --editable providers/imap --editable providers/exasol --editable providers/pgvector --editable providers/atlassian/jira --editable providers/google --editable providers/jenkins --editable providers/teradata --editable providers/segment --editable providers/salesforce --editable providers/influxdb --editable providers/redis --editable providers/apache/hive --editable providers/apache/pinot --editable providers/apache/livy --editable providers/apache/drill --editable providers/apache/kafka --editable providers/apache/druid --editable providers/apache/pig --editable providers/apache/cassandra --editable providers/apache/kylin --editable providers/apache/iceberg --editable providers/apache/spark --editable providers/apache/beam --editable providers/mongo --editable providers --editable providers/alibaba --editable providers/mysql --editable providers/sftp --editable providers/celery --editable providers/datadog --editable providers/trino --editable providers/ssh --editable providers/airbyte --editable providers/openlineage --editable providers/github --editable providers/ftp --editable providers/ydb --editable providers/opsgenie --editable providers/weaviate --editable providers/microsoft/psrp --editable providers/microsoft/mssql --editable providers/oracle --editable providers/postgres --editable providers/fab --editable providers/telegram --editable providers/odbc --editable providers/samba --editable providers/qdrant --editable providers/papermill --editable providers/sendgrid --editable providers/zendesk --editable providers/openai --editable providers/facebook --editable providers/http --editable providers/openfaas --editable providers/dbt/cloud --editable providers/hashicorp --editable providers/edge --editable providers/tableau --editable providers/presto --editable providers/discord --editable providers/neo4j --editable providers/sqlite --editable providers/dingding --editable providers/slack --editable providers/jdbc --editable providers/opensearch --editable providers/apprise
  #45 1.035 Using Python 3.9.21 environment at: /usr/local
  #45 3.203   × No solution found when resolving dependencies:
  #45 3.203   ╰─▶ Because the current Python version (3.9.21) does not satisfy
  #45 3.203       Python>=3.10,<4 and apache-airflow-providers-cloudant==4.1.0
  #45 3.203       depends on Python>=3.10,<4, we can conclude that
  #45 3.203       apache-airflow-providers-cloudant==4.1.0 cannot be used.
  #45 3.203       And because only apache-airflow-providers-cloudant==4.1.0 is available
  #45 3.203       and you require apache-airflow-providers-cloudant, we can conclude that
  #45 3.203       your requirements are unsatisfiable.
  #45 ERROR: process "/bin/bash -o pipefail -o errexit -o nounset -o nolog -c bash /scripts/docker/install_airflow.sh" did not complete successfully: exit code: 1
  ------
   > [main 16/24] RUN --mount=type=cache,id=ci-amd64-0,target=/root/.cache/ bash /scripts/docker/install_airflow.sh:
  0.956 + uv pip install --system --upgrade --resolution highest --editable '.[devel-ci]' --editable ./task_sdk --editable providers/pagerduty --editable providers/common/compat --editable providers/common/io --editable providers/common/sql --editable providers/cloudant --editable providers/asana --editable providers/pinecone --editable providers/cohere --editable providers/grpc --editable providers/singularity --editable providers/elasticsearch --editable providers/docker --editable providers/vertica --editable providers/snowflake --editable providers/standard --editable providers/imap --editable providers/exasol --editable providers/pgvector --editable providers/atlassian/jira --editable providers/google --editable providers/jenkins --editable providers/teradata --editable providers/segment --editable providers/salesforce --editable providers/influxdb --editable providers/redis --editable providers/apache/hive --editable providers/apache/pinot --editable providers/apache/livy --editable providers/apache/drill --editable providers/apache/kafka --editable providers/apache/druid --editable providers/apache/pig --editable providers/apache/cassandra --editable providers/apache/kylin --editable providers/apache/iceberg --editable providers/apache/spark --editable providers/apache/beam --editable providers/mongo --editable providers --editable providers/alibaba --editable providers/mysql --editable providers/sftp --editable providers/celery --editable providers/datadog --editable providers/trino --editable providers/ssh --editable providers/airbyte --editable providers/openlineage --editable providers/github --editable providers/ftp --editable providers/ydb --editable providers/opsgenie --editable providers/weaviate --editable providers/microsoft/psrp --editable providers/microsoft/mssql --editable providers/oracle --editable providers/postgres --editable providers/fab --editable providers/telegram --editable providers/odbc --editable providers/samba --editable providers/qdrant --editable providers/papermill --editable providers/sendgrid --editable providers/zendesk --editable providers/openai --editable providers/facebook --editable providers/http --editable providers/openfaas --editable providers/dbt/cloud --editable providers/hashicorp --editable providers/edge --editable providers/tableau --editable providers/presto --editable providers/discord --editable providers/neo4j --editable providers/sqlite --editable providers/dingding --editable providers/slack --editable providers/jdbc --editable providers/opensearch --editable providers/apprise
  1.035 Using Python 3.9.21 environment at: /usr/local
  3.203   × No solution found when resolving dependencies:
  3.203   ╰─▶ Because the current Python version (3.9.21) does not satisfy
  3.203       Python>=3.10,<4 and apache-airflow-providers-cloudant==4.1.0
  3.203       depends on Python>=3.10,<4, we can conclude that
  3.203       apache-airflow-providers-cloudant==4.1.0 cannot be used.
  3.203       And because only apache-airflow-providers-cloudant==4.1.0 is available
  3.203       and you require apache-airflow-providers-cloudant, we can conclude that
  3.203       your requirements are unsatisfiable.
  ------
  Dockerfile.ci:1308
  --------------------
   1306 |     # But in cron job we will install latest versions matching pyproject.toml to see if there is no breaking change
   1307 |     # and push the constraints if everything is successful
   1308 | >>> RUN --mount=type=cache,id=ci-$TARGETARCH-$DEPENDENCY_CACHE_EPOCH,target=/root/.cache/ bash /scripts/docker/install_airflow.sh
   1309 |     
   1310 |     COPY --from=scripts install_packaging_tools.sh install_additional_dependencies.sh /scripts/docker/
  --------------------
  ERROR: failed to solve: process "/bin/bash -o pipefail -o errexit -o nounset -o nolog -c bash /scripts/docker/install_airflow.sh" did not complete successfully: exit code: 1
Error when building image! Image build: 3.9:linux/amd64
Error: Process completed with exit code 1.
```

o-nikolas replied:
From what I can tell since `!=3.9` is not a valid exclusion (because there is no operator expanding it and `3.9` on its own is not a valid python version) we were actually allowing cloudant to be installed in 3.9 builds... So this was always kind of broken until now.

o-nikolas replied:
I ended up adding a case for cloudant in `install_airflow.sh`, if we are on python3.9 exclude that provider from the build.

o-nikolas replied:
That got me past image build failures, but now several tests downstream use that image in which cloudant is not longer installed, so many of the tests are failing (the static check to verify hooks for example is detecting that we should have a provider cloudant that has hooks but that module isn't found in the image because it wasn't installed).

potiuk replied:
Yeah, I think it's indeed a red-herring we chase. I believe we should leave ~3.9 in all providers and stay with "soft" exclusion with "trove classifiers". The main problem is that I think this is a `uv workspace` limitaiton that all projects that are part of the workspace have to have the same python version support:

From https://docs.astral.sh/uv/concepts/projects/workspaces/#when-not-to-use-workspaces

> Finally, uv's workspaces enforce a single requires-python for the entire workspace, taking the intersection of all members' requires-python values. If you need to support testing a given member on a Python version that isn't supported by the rest of the workspace, you may need to use uv pip to install that member in a separate virtual environment.

So this is likely not something that we can workaround today - and "trove classifiers" and the explanation in the README are good enough of a signal to tell people which versions of python are supported per provider. I don't think we can do better than that.

potiuk replied:
But would be nice to at least generate `~3.9` from the minimum Python version :) . That will not change generated pyproject.toml but at least all that effort here will not be completely lost :D

o-nikolas replied:
Okay, I'll revert back to just adding a "broken" (could be interpreted as docs only) exclusion for 3.9 (i.e. back to "~=3.9,!=3.9") this will signal to the user that 3.9 is not to be used, and will update the trove classifiers, but won't actually affect dependency resolution.

potiuk commented on dev/breeze/src/airflow_breeze/prepare_providers/provider_documentation.py:
```diff
@@ -1222,6 +1222,13 @@ def _regenerate_pyproject_toml(context: dict[str, Any], provider_details: Provid
     context["DEPENDENCY_GROUPS"] = "\n".join(dependency_groups)
 
+    # Most providers require the same python versions, but some may have exclusions
+    requires_python_version = "~=3.9"
```
NIT: we coud use `DEFAULT_PYTHON_MAJOR_MINOR_VERSION` from global_constants.py here. This way we will not have to change it when we bump to 3.10.

o-nikolas replied:
Yupp, had this thought as I went to bed last night, I didn't like this magic string here in the middle of the code. I'll update this

potiuk commented on dev/breeze/src/airflow_breeze/prepare_providers/provider_documentation.py:
```diff
@@ -1222,6 +1222,13 @@ def _regenerate_pyproject_toml(context: dict[str, Any], provider_details: Provid
+    # Most providers require the same python versions, but some may have exclusions
+    requires_python_version = "~=3.9"
+    for excluded_python_version in provider_details.excluded_python_versions:
+        requires_python_version += f", !={excluded_python_version}"
```
NIT: In most other places we generate version spec without spaces. While it's ok to have them and specs explicitly has spaces, all whitespace is normalized and removed anyway https://packaging.python.org/en/latest/specifications/version-specifiers/#version-specifiers  

That's more of an easthetic thing - but I think it would be good if we keep consistency here with other dependencies - for example in celery`:

```python
# The dependencies should be modified in place in the generated file
# Any change in the dependencies is preserved when the file is regenerated
dependencies = [
    "apache-airflow>=2.9.0",
    # The Celery is known to introduce problems when upgraded to a MAJOR version. Airflow Core
    # Uses Celery for CeleryExecutor, and we also know that Kubernetes Python client follows SemVer
    # (https://docs.celeryq.dev/en/stable/contributing.html?highlight=semver#versions).
    "celery[redis]>=5.4.0,<6",
    "flower>=1.0.0",
    "google-re2>=1.0",
]
```

I personally prefer to avoid whitespace in such cases - it decreases the chance that somewhere in the code where space is treated as separator and you forgot to quote things the dependency will be mistakenly split into several separate specs.

o-nikolas replied:
Sure, I saw spaces and not spaces in all the things I read but mostly spaces in docs. I don't have a preference, happy to remove the space.

o-nikolas replied:
Ah yes, I remember now, ruff fails when I remove the space. It has a hard time parsing the version specifier when it's omitted:
```
ruff failed
  Cause: Failed to parse /home/ANT.AMAZON.COM/onikolas/code/oss/airflow/team_fork/providers/cloudant/pyproject.toml
  Cause: TOML parse error at line 52, column 19
   |
52 | requires-python = "3.9,!=3.9"
   |                   ^^^^^^^^^^^
Failed to parse version: Unexpected end of version specifier, expected operator:
3.9,!=3.9
^^^
```

o-nikolas replied:
Actually, this is a different bug that was introduced at the same time. Misting the `~=` operator

potiuk commented on providers/cloudant/pyproject.toml:
```diff
@@ -0,0 +1,79 @@
+    "apache-airflow>=2.9.0",
+    # Even though 3.9 is excluded below, we need to make this python_version aware so that `uv` can generate a
+    # full lock file when building lock file from provider sources
+    "ibmcloudant==0.9.1 ; python_version >= \"3.10\"",
```
NIT: Same here -> I prefer spaces to be removed :)

o-nikolas replied:
This one had spaces from whoever added it before and it was just copied by the script. But I can update that now after the script is done.

potiuk commented on providers/cloudant/src/airflow/providers/cloudant/get_provider_info.py:
```diff
@@ -0,0 +1,78 @@
+            }
+        ],
+        "dependencies": ["apache-airflow>=2.9.0", 'ibmcloudant==0.9.1 ; python_version >= "3.10"'],
+    }
```
NIT: maybe we can add "requires_python" here as well. The ``get_provider_info` can have new fields, no problem with that - and it would be nice to specify it here

o-nikolas replied:
Sure, I don't yet grok why we need all this info in two places, but I don't mind adding it. Should we do that in a separate PR since it will affect all providers?

potiuk replied:
> Sure, I don't yet grok why we need all this info in two places, but I don't mind adding it. Should we do that in a separate PR since it will affect all providers?

Yes.

potiuk commented on providers/cloudant/README.rst:
```diff
@@ -0,0 +1,62 @@
+``apache-airflow``  ``>=2.9.0``
+``ibmcloudant``     ``==0.9.1; python_version >= "3.10"``
+==================  =====================================
+
```
NIT: An idea - this README is something immediatly visible for package in PyPI -> maybe we could also add information about Python versions supported?

o-nikolas replied:
That could be a neat follow up. But it will affect all providers, perhaps we update it in a separate PR?

potiuk replied:
Actually... I am totally stupid... It's already up there :) :+1: 

>  The package supports the following python versions: 3.10,3.11,3.12

potiuk commented on dev/breeze/src/airflow_breeze/prepare_providers/provider_documentation.py:
```diff
@@ -1222,6 +1222,13 @@ def _regenerate_pyproject_toml(context: dict[str, Any], provider_details: Provid
     context["EXTRAS_REQUIREMENTS"] = "\n".join(optional_dependencies)
     context["DEPENDENCY_GROUPS"] = "\n".join(dependency_groups)
 
+    # Most providers require the same python versions, but some may have exclusions
```
That's not the right place to set the context values (that's why the CI fails) - here we only override the values that we manually modify in pyproject.toml  (dependencies are manually modified directly in pyproject.toml - so before we regenerate pyproject.toml we read them from there - including comments - so that they are used when we regenerate pyproject.toml with other changes. 

The right place where to set the context is `get_provider_jinja_context` in  `src/airflow_breeze/utils/packages.py`

o-nikolas replied:
By context do you mean this comment itself? Or the `context` object above?

I didn't intend for this comment to show in the pyproject.toml, which is why I put it here. It was more to explain the code itself. But I can also add one to the toml template if you think that's useful context to have in there?

o-nikolas replied:
Ah, I see you mean the context object itself. It's used in several places, makes sense. I'll move the below logic over to `get_provider_jinja_context`

o-nikolas committed:
Updates from PR feedback

o-nikolas committed:
Add wildcard for all patch releases of an excluded python version

If it is not already present add a wild card for all patch releases of
an excluded python version

o-nikolas committed:
Fix install_airflow script to remove cloudant on 3.9

o-nikolas committed:
Remove more spaces

o-nikolas committed:
Remove the fix for version exclusions

With functional version exclusion in requires-python working, many
things in our CI are not setup to support this. So go back to "soft"
or non-functional exclusions. This will at least update the classifiers
